File: geeks_for_geeks/remove_K_digits.py
class Solution:
    def removeKdigits(self, S, K):
        n = len(S)
        if K==n-1:
            return min(S)
        if K==n:
            return 0
        ret = list() ; popped = 0 ; cnt = 1
        idx = 1 ; ret.append(S[0])
        while popped < K and idx<n:
            if int(ret[-1]) > int(S[idx]):
                while ret and idx<n and popped<K and (int(ret[-1]) >= int(S[idx])):
                    ret.pop(-1); popped += 1
                    cnt -= 1
                ret.append(S[idx]) ; idx += 1
                cnt += 1
            else:
                ret.append(S[idx])
                idx += 1 ; cnt += 1
        while cnt<n-K and idx<n:
            ret.append(S[idx])
            idx += 1
        ret = ret[:n-K]
        return ''.join(ret)

File: geeks_for_geeks/remove_K_Digits.java
class Solution {
    public String removeKdigits(String S, int K) {
        int n = S.length(), popped = 0, idx = 1, i=0;
        int[] digits = new int[S.length()];
        for(i=0; i<S.length(); i++)
            digits[i] = S.charAt(i)-'0';
        if (K==n-1) return Integer.toString(Arrays.stream(digits).min().getAsInt());
        if (K==n) return "0";
        Stack<Integer> stack = new Stack<Integer>();
        stack.push(S.charAt(0)-'0');
        while (idx < n)
        {
            while (!stack.isEmpty() && popped<K && (stack.peek()  > S.charAt(idx)-'0' ) )
            {
                stack.pop(); popped += 1;
            }
            stack.push(S.charAt(idx)-'0'); idx += 1;
        }
        int size = stack.size();
        for(i=idx; i<n; i++)
            stack.push(S.charAt(i)-'0');
        StringBuilder ret = new StringBuilder();
        while (!stack.isEmpty())
            ret.append(stack.pop());
        String retStr = ret.reverse().toString().substring(0,n-K);
        for (i=0; i<n-K; i++)
            if (retStr.charAt(i)!='0') break;
        if (i==n-K) return "0";
        return retStr.substring(i,n-K);
    }
}